### PR TITLE
change: add a 1 second sleep for List operations when a LimitExceededException is thrown

### PR DESCRIPTION
**Pull request by yzhu0**
Add throttle to list methods

### CODE REVIEW

**Comment by icywang86rui** [comment]
I am not sure this is the correct way to fix throttling problems. This will add 1 second delay in all the ListApi calls wherever this is used. This might break our customers that has a higher rate for their specific account.

**Review by icywang86rui — CHANGES_REQUESTED** [review]
Not sure this change is the right fix.

**Review by jerrypeng7773 — DISMISSED** [review]

**Comment by danabens** [comment]
fix unit tests
`[Container] 2021/01/06 18:45:23 Phase context status code: COMMAND_EXECUTION_ERROR Message: Error while executing command: tox -e flake8,pylint,docstyle,black-check,twine --parallel all. Reason: exit status 1`

**Comment by yzhu0** [comment]
> I am not sure this is the correct way to fix throttling problems. This will add 1 second delay in all the ListApi calls wherever this is used. This might break our customers that has a higher rate for their specific account.

since we have rate =2 under ListApi call. We should let it sleep for 1s to prevent API throttling
https://tiny.amazon.com/xh7mx9uw/IronmanUtilizationService/342d6fd1/default-rules.json#L225

**Review by danabens — APPROVED** [review]

**Review by danabens — COMMENTED** [review]
Rui's comments are correct - should normally list at full speed, once there is a throttling exception list at a slower rate.

**Comment by yzhu0** [comment]
> Rui's comments are correct - should normally list at full speed, once there is a throttling exception list at a slower rate.

I update the pr to sleep only when the method is throttled

**Comment by danabens** [comment]
confirm tested end to end either via integ test or manually

File: src/sagemaker/apiutils/_base_types.py [hunk 1]
```diff
@@ -130,19 +132,28 @@ def _list(
...
+                    if not next_token:
+                        break
+                except botocore.exceptions.ClientError as error:
+                    if error.response["Error"]["Code"] == "LimitExceededException":
+                        logging.getLogger().warning(
+                            "API call limit exceeded; backing off and retrying..."
+                        )
+                        time.sleep(1)
+                    else:
+                        raise error
```
Comment (ChoiByungWook) [inline src/sagemaker/apiutils/_base_types.py:156]: Can we update unit tests to exercise this line?

**Review by ChoiByungWook — CHANGES_REQUESTED** [review]

**Review by ChoiByungWook — COMMENTED** [review]
Would it be better to configure the retry strategy in the boto client instead of sleeping?

https://boto3.amazonaws.com/v1/documentation/api/latest/guide/retries.html

**Comment by icywang86rui** [comment]
There has not been any actions for a few weeks. Closing for now. Feel free to reopen.

**Pull request closed without merging**